storage: experimental Raft SSTable sideloading (in-memory storage)

tbg opened the pull request:
Ready for review:

- [x] determine whether it's worth jumping through hoops to move some of this
  code to CCL (testing complicated unless we soften the boundary some).
- [x] tests for maybeInline
- [x] fix up some locks
- [x] tests for maybeSideload
- [x] tests for replicaGC
- [x] tests for log truncation
- [x] test for snapshot's constructed Raft log entries (must be inlined)
- [x] make naming more consistent: currently there is fat, thin, sideloaded,
  SSTable, payload, ...

Next PRs (see https://github.com/cockroachdb/cockroach/issues/16263) will adress the left-over work:

- implement actual on-disk storage 
  - decide on/implement actual storage location
  - clean up `addSSTablePostApply`
- performance counters, stats & visibility into stored files


----

Implement the Raft SSTable Sideloading RFC. Briefly put,

- on the Replica proposal path, detect AddSSTable commands and mark them as
  Raft command version v2.
- Intercept such entries on their way into the Raft log, and store the
  actual SSTable on disk.
- On the wire, always send the full ("inlined") proposals, including in
  snapshots.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16509)
<!-- Reviewable:end -->

tschottdorf commented:
Ok, this is now open for review. I think it's reasonable to address both the actual on-disk storage (currently uses an in-memory map) and diagnostics as a follow-up. I want @danhhz to try this out and this PR is already large enough.

danhhz commented:
You've got some merge conflicts, so you probably want to pull and resolve them. That will also get you the updated benchmarks, which we should run and include in the commit message

---

Review status: 0 of 17 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/ccl/storageccl/import.go, line 45 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnAfYujsppwVtm533Cd:-KnAfYuk_XtGYtzCsk2W:b-76mjtk) ([raw file](https://github.com/cockroachdb/cockroach/blob/74c366ca230536b8c4c9dcb207374c6d7c7bb8a6/pkg/ccl/storageccl/import.go#L45)):*
> ```Go
> func init() {
> 	storage.SetImportCmd(evalImport)
> 	storage.CanSideloadSSTable = AddSSTableEnabled.Get
> ```

This is not quite the right thing to use. `AddSSTableEnabled` controls whether the Import command uses WriteBatch or AddSSTable. If we're trying to enterprise gate it, you really want something like

```
storage.CanSideloadSSTable = func() bool {
  err := utilccl.CheckEnterpriseEnabled("AddSSTable")
  return err == nil
}
```

which should probably live in addsstable.go

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-:-KnAgGNZKLSLI5uLdkmK:b-bj6fy1)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 11 of 17 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/storageccl/import.go, line 45 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnAfYujsppwVtm533Cd:-KnBBmXkMoBI3DdZpi9j:b-mcq8x9) ([raw file](https://github.com/cockroachdb/cockroach/blob/74c366ca230536b8c4c9dcb207374c6d7c7bb8a6/pkg/ccl/storageccl/import.go#L45)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

This is not quite the right thing to use. `AddSSTableEnabled` controls whether the Import command uses WriteBatch or AddSSTable. If we're trying to enterprise gate it, you really want something like

```
storage.CanSideloadSSTable = func() bool {
  err := utilccl.CheckEnterpriseEnabled("AddSSTable")
  return err == nil
}
```

which should probably live in addsstable.go
</blockquote></details>

We don't want to enterprise-gate things in the non-ccl parts of the repo. Unless we can arrange to have the code live on the ccl side, just leave it open. I would rename the setting to something like `CanProposeSideloadedCommand` since the actual sideloading behavior is always enabled. (Maybe the setting itself should live on the OSS side?)

---

*[pkg/storage/replica.go, line 3210 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBHRbj9HsB9GuOBlpx:-KnBHRbj9HsB9GuOBlpy:b-q4jk1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica.go#L3210)):*
> ```Go
> 	//
> 	// Note also that here we cache any sideloaded SSTables that were just
> 	// inlined above. This is very relevant as we're likely to apply them
> ```

Doesn't look like anything's actually inlined above. 

---

*[pkg/storage/replica.go, line 3234 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBHLDwvnfhKZtagX6s:-KnBHLDwvnfhKZtagX6t:bp0cshh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica.go#L3234)):*
> ```Go
> 					// in transit anyway, but there is likely storage layer corruption
> 					// going on, something we realistically don't handle at the moment.
> 					log.Fatalf(ctx, "while inlining sideloaded commands: %s", err)
> ```

I would still make this a log.Error and drop the message instead of killing the process.

---

*[pkg/storage/replica_proposal.go, line 508 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBHog54cOB-EumdODZ:-KnBHog54cOB-EumdOD_:brbbdjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_proposal.go#L508)):*
> ```Go
> 		}
> 	} else {
> 		// TODO(danhhz,tschottdorf): we don't want a temp dir here. Need a concrete
> ```

Is this comment still valid with the GetAuxiliaryDir change?

---

*[pkg/storage/replica_proposal.go, line 687 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBHwXfiA5vtFmkYDwK:-KnBHwXfiA5vtFmkYDwL:b-u81odw) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_proposal.go#L687)):*
> ```Go
> 		// to and including the most recently truncated index.
> 		r.store.raftEntryCache.clearTo(r.RangeID, newTruncState.Index+1)
> 		log.Eventf(ctx, "truncating log up to (and including) index %d", newTruncState.Index)
> ```

s/log/sideloaded storage/

---

*[pkg/storage/replica_proposal.go, line 692 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBI47sT7XP8bprTIeQ:-KnBI47sT7XP8bprTIeR:b-95hwbm) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_proposal.go#L692)):*
> ```Go
> 			// is unexpected except when the storage medium fails, and we don't handle that
> 			// graceful anywhere else.
> 			log.Fatalf(ctx, "while removing sideloaded files during log truncation: %s", err)
> ```

Again, I'd log an error and continue when we know we can do so safely. 

---

*[pkg/storage/replica_raftstorage.go, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBI_FuhcNg8QMmWdgJ:-KnBI_FuhcNg8QMmWdgK:b-9dcq2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_raftstorage.go#L80)):*
> ```Go
> // maxBytes. Passing maxBytes equal to zero disables size checking. Sideloaded
> // files don't count towards maxBytes; only the part of the entry that lives in
> // the Raft log does.
> ```

Is that true? I don't see why it would be. And I think we want to count sideloaded data towards maxBytes; otherwise the rehydrated entries could consume much more memory than we expect.

---

*[pkg/storage/replica_raftstorage.go, line 511 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBKNw3MkhdC_LhE6SY:-KnBKNw3MkhdC_LhE6SZ:btv6zuu) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_raftstorage.go#L511)):*
> ```Go
> 	ctx context.Context,
> 	batch engine.ReadWriter,
> 	sideloaded sideloadStorage,
> ```

This argument looks unused.

---

*[pkg/storage/replica_raftstorage.go, line 543 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBKGVrhGIVbcUhm_-D:-KnBKGVrhGIVbcUhm_-E:b4xvwgi) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_raftstorage.go#L543)):*
> ```Go
> 	lastIndex := entries[len(entries)-1].Index
> 	for i := lastIndex + 1; i <= prevLastIndex; i++ {
> 		// TODO(tschottdorf): can also delete any corresponding sideloaded
> ```

Isn't this done above?

---

*[pkg/storage/replica_raftstorage.go, line 819 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBKS5q5KXxl0eL6u1k:-KnBKS5q5KXxl0eL6u1l:b-r8uryt) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_raftstorage.go#L819)):*
> ```Go
> 	raftCommandIDLen = 8
> 	// The initial Raft command version, used for all regular Raft traffic.
> 	raftCommandEncodingVersionV1 raftCommandEncodingVersion = 0
> ```

Use V0 and V1 in the constant and function names to match the values. (or give them more meaningful names like raftCommandEncodingSideload)

---

*[pkg/storage/replica_sideload.go, line 45 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBPZ_rBM0ZnPkIH-tg:-KnBPZ_rBM0ZnPkIH-th:bfkhef5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload.go#L45)):*
> ```Go
> 	// exists, returns errSideloadedFileNotFound.
> 	Purge(_ context.Context, index, term uint64) error
> 	// Clear removes all on-disk files.
> ```

s/files/files that may have been written by this sideloadStorage/

---

*[pkg/storage/replica_sideload.go, line 138 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBZNmt0fZmCSKLAuAG:-KnBZNmuxhmnL8aLxqBm:b-nz0lx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload.go#L138)):*
> ```Go
> 		cmd, ok := r.mu.proposals[cmdID]
> 		if !ok {
> 			return storagebase.RaftCommand{}
> ```

I'd prefer a `, ok` return type instead of returning an empty value that must be inspected.

---

*[pkg/storage/replica_sideload.go, line 152 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnB_kK_q6vOQvP_3J5X:-KnB_kK_q6vOQvP_3J5Y:b-d90pz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload.go#L152)):*
> ```Go
> 		// to modify.
> 		rCmd := cmd.command
> 		addSSTableCopy := *cmd.command.ReplicatedEvalResult.AddSSTable
> ```

I think it would be cleaner for the caller to make this copy. 

---

*[pkg/storage/replica_sideload.go, line 163 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBbfRAdpdcCQGM2MEO:-KnBbfRAdpdcCQGM2MEP:b958mjg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload.go#L163)):*
> ```Go
> }
> 
> func maybeInlineSideloadedRaftCommand(
> ```

Document the error behavior (and specifically that it returns an error whose *cause* is `errSideloadedFileNotFound`), since we rely on it elsewhere.

---

*[pkg/storage/replica_sideload.go, line 214 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBaJoOHbgDZOwOjk_l:-KnBaJoOHbgDZOwOjk_m:b-ndolul) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload.go#L214)):*
> ```Go
> // returns a new slice in which all applicable entries have been sideloaded to
> // the specified sideloadStorage. Whenever maybeRaftCommand returns a Raft
> // command, it is assumed that its AddSSTable field may be mutated.
> ```

This last sentence should also describe what maybeRaftCommand is/does. This is an awkward place to split the functionality. Maybe the argument should be a possibly-nil pointer to the `r.mu.proposals` map, since that's the only implementation that matters?

---

*[pkg/storage/replica_sideload.go, line 215 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBaQxOo-2LugCJygH1:-KnBaQxPQxKPfYUI1cAs:b-jdjde7) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload.go#L215)):*
> ```Go
> // the specified sideloadStorage. Whenever maybeRaftCommand returns a Raft
> // command, it is assumed that its AddSSTable field may be mutated.
> func maybeSideloadEntriesImpl(
> ```

Move this up next to maybeSideloadEntriesRaftMuLocked. 

---

*[pkg/storage/replica_sideload_test.go, line 48 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBdLTIjVt7eGFiL873:-KnBdLTJ69e21HRFkiu4:b-uwf17i) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload_test.go#L48)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 
> 	t.Run("inline", testSideloadInline)
> ```

Why do these as subtests? It's an unusual pattern when there's nothing shared between them (and it means that in the event that leaktest fails, we can't tell which subtest is responsible)

---

*[pkg/storage/replica_sideload_test.go, line 56 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBdnT38oAFvpitIAd9:-KnBdnT38oAFvpitIAdA:b-jtszcy) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload_test.go#L56)):*
> ```Go
> }
> 
> func eq(l, r raftpb.Entry) error {
> ```

This name has an unfortunately broad scope (all the tests in package `storage`). Probably better to give it a longer name (and make a function-local abbreviation if you need to)

---

*[pkg/storage/replica_sideload_test.go, line 155 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBetgz88IDsLGkF-Y2:-KnBetgz88IDsLGkF-Y3:b-agqa3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload_test.go#L155)):*
> ```Go
> 		"v2-with-payload-with-file-no-cache": {
> 			fat: mkEnt(v2, 5, 6, &sstThin), thin: mkEnt(v2, 5, 6, &sstFat),
> 			setup: putOnDisk, expTrace: "inlined entry not cached",
> ```

Nice use of tracing to ensure coverage of all the important cases.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/storageccl/import.go, line 45 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnAfYujsppwVtm533Cd:-Kn_TVDNX3l_xIqqLP8v:b-1pyd9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/74c366ca230536b8c4c9dcb207374c6d7c7bb8a6/pkg/ccl/storageccl/import.go#L45)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We don't want to enterprise-gate things in the non-ccl parts of the repo. Unless we can arrange to have the code live on the ccl side, just leave it open. I would rename the setting to something like `CanProposeSideloadedCommand` since the actual sideloading behavior is always enabled. (Maybe the setting itself should live on the OSS side?)
</blockquote></details>

Removed `CanSideloadSSTable`.

---

*[pkg/storage/replica.go, line 3210 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBHRbj9HsB9GuOBlpx:-KnZtknpZlUPPMCMYsFf:b-dwiyr5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica.go#L3210)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Doesn't look like anything's actually inlined above. 
</blockquote></details>

Fixed the comment.

---

*[pkg/storage/replica.go, line 3234 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBHLDwvnfhKZtagX6s:-KnZtvXXV4lTnKTIqfZ2:b-lj8afi) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica.go#L3234)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I would still make this a log.Error and drop the message instead of killing the process.
</blockquote></details>

I don't really agree with this, but made the change.

---

*[pkg/storage/replica_proposal.go, line 508 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBHog54cOB-EumdODZ:-KnZvC8AGHPRJRapMk2O:b-yz69p2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_proposal.go#L508)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Is this comment still valid with the GetAuxiliaryDir change?
</blockquote></details>

Updated. Also made this idempotent (i.e. if hashed file already exists, noop). Probably want something better here before we roll this out, but that should be done with https://github.com/cockroachdb/cockroach/pull/16633.

---

*[pkg/storage/replica_proposal.go, line 687 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBHwXfiA5vtFmkYDwK:-KnZvYbJidRcwM1tQTAT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_proposal.go#L687)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/log/sideloaded storage/
</blockquote></details>

Done.

---

*[pkg/storage/replica_proposal.go, line 692 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBI47sT7XP8bprTIeQ:-KnZvZb2vR0Ph6a6QXvB:b-rqxvo7) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_proposal.go#L692)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Again, I'd log an error and continue when we know we can do so safely. 
</blockquote></details>

ditto.

---

*[pkg/storage/replica_raftstorage.go, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBI_FuhcNg8QMmWdgJ:-Kn_QL1lKLNFRO2CzQgb:blmfqdx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_raftstorage.go#L80)):*
> Is that true? I don't see why it would be.

It is now (with the latest diff). Opens a potential can of worms - we need to make sure to keep accurate enough stats to not confuse the `raftLogSize` (which in turn influences log truncation). Haven't done so yet, but added TODOs in the relevant places.

---

*[pkg/storage/replica_raftstorage.go, line 511 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBKNw3MkhdC_LhE6SY:-Kn_-0CHe9oruGuMvwDn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_raftstorage.go#L511)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This argument looks unused.
</blockquote></details>

Done.

---

*[pkg/storage/replica_raftstorage.go, line 543 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBKGVrhGIVbcUhm_-D:-Kn_-YzFvIFkKC5pGjEZ:b-i9rr6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_raftstorage.go#L543)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Isn't this done above?
</blockquote></details>

Yep, removed this and added commentary on the function.

---

*[pkg/storage/replica_raftstorage.go, line 819 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBKS5q5KXxl0eL6u1k:-Kn_JxUL93yJV5wyT3mL:b-f8hxdc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_raftstorage.go#L819)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Use V0 and V1 in the constant and function names to match the values. (or give them more meaningful names like raftCommandEncodingSideload)
</blockquote></details>

Good point, now `raftVersion{Standard,Sideloaded}`.

---

*[pkg/storage/replica_sideload.go, line 45 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBPZ_rBM0ZnPkIH-tg:-Kn_K9tvQW-sFZuRJjum:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload.go#L45)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/files/files that may have been written by this sideloadStorage/
</blockquote></details>

Done.

---

*[pkg/storage/replica_sideload.go, line 138 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBZNmt0fZmCSKLAuAG:-Kn_LQXB1rCzDkzcDKiD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload.go#L138)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'd prefer a `, ok` return type instead of returning an empty value that must be inspected.
</blockquote></details>

Done.

---

*[pkg/storage/replica_sideload.go, line 152 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnB_kK_q6vOQvP_3J5X:-Kn_LQwl_Tp3sI6oDsvC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload.go#L152)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think it would be cleaner for the caller to make this copy. 
</blockquote></details>

Done.

---

*[pkg/storage/replica_sideload.go, line 163 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBbfRAdpdcCQGM2MEO:-Kn_MBpEve1IC3R9eRyu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload.go#L163)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Document the error behavior (and specifically that it returns an error whose *cause* is `errSideloadedFileNotFound`), since we rely on it elsewhere.
</blockquote></details>

Done.

---

*[pkg/storage/replica_sideload.go, line 214 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBaJoOHbgDZOwOjk_l:-Kn_NL77NFTVKjYGzGDk:b-5yd1uz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload.go#L214)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This last sentence should also describe what maybeRaftCommand is/does. This is an awkward place to split the functionality. Maybe the argument should be a possibly-nil pointer to the `r.mu.proposals` map, since that's the only implementation that matters?
</blockquote></details>

I thought about that initially, but ended up siding with the more modular API. `maybeRaftCommand` is now fairly trivial (since it doesn't do the copy any more), so it looks good to me.

---

*[pkg/storage/replica_sideload.go, line 215 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBaQxOo-2LugCJygH1:-Kn_NKuLtFk_GCqayDOb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload.go#L215)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Move this up next to maybeSideloadEntriesRaftMuLocked. 
</blockquote></details>

Done.

---

*[pkg/storage/replica_sideload_test.go, line 48 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBdLTIjVt7eGFiL873:-Kn_NvcFbWYryD5zqEvb:b-nwddwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload_test.go#L48)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why do these as subtests? It's an unusual pattern when there's nothing shared between them (and it means that in the event that leaktest fails, we can't tell which subtest is responsible)
</blockquote></details>

because it's nice to be able to run `TESTS=Sideloading` and get the sideloading-related tests. But maybe I went overboard with this. The leaktest thing is unfortunate, but more a limitation of leaktest than indicating that this isn't a good pattern. Happy to make these individual tests if you think that's what they should be.

---

*[pkg/storage/replica_sideload_test.go, line 56 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBdnT38oAFvpitIAd9:-Kn_QEc7CZYFcKcAiXEM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload_test.go#L56)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This name has an unfortunately broad scope (all the tests in package `storage`). Probably better to give it a longer name (and make a function-local abbreviation if you need to)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 7 of 17 files reviewed at latest revision, 18 unresolved discussions.

---

*[pkg/storage/replica_raftstorage.go, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBI_FuhcNg8QMmWdgJ:-Knf9hJZxbdiYHypj6pI:bldaid9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_raftstorage.go#L80)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

> Is that true? I don't see why it would be.

It is now (with the latest diff). Opens a potential can of worms - we need to make sure to keep accurate enough stats to not confuse the `raftLogSize` (which in turn influences log truncation). Haven't done so yet, but added TODOs in the relevant places.
</blockquote></details>

Disregard the "can of worms" part in my previous comment; I took your comment to mean that the entries should count towards `raftLogSize` and that's hard, but I don't think that makes sense nor is it what you wanted. `entries()` should now correctly count the rehydrated bytes, though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 10 of 10 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_proposal.go, line 520 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KngiUQxwO1qAQ6mhxAm:-KngiUQxwO1qAQ6mhxAn:b-d6gbz2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e5ca533aceaa64a642dc1ce2644b3642ba779b8/pkg/storage/replica_proposal.go#L520)):*
> ```Go
> 			panic(err)
> 		} else {
> 			log.Warningf(ctx, "file %s already exists, likely doubly applied Raft command; skipping", path)
> ```

The file could have been incompletely written by an earlier version of the process. We should verify that the existing file has the contents we want to write. 

---

*[pkg/storage/replica_sideload_test.go, line 48 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBdLTIjVt7eGFiL873:-KnghYyupykmPXZIc1un:b-b8kbcr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload_test.go#L48)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

because it's nice to be able to run `TESTS=Sideloading` and get the sideloading-related tests. But maybe I went overboard with this. The leaktest thing is unfortunate, but more a limitation of leaktest than indicating that this isn't a good pattern. Happy to make these individual tests if you think that's what they should be.
</blockquote></details>

`TESTS=Sideloading` does a regex search on the test names, so as long as all the tests have `Sideloading` in their names it will work the way you want. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-:-KngjLuqrE1oDK5tL2fs:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR Ben. Just FYI, using this on lapis for a TPC restore quickly causes
inconsistency panics. I hope to track down why tomorrow.

On Tue, Jun 27, 2017, 23:27 Ben Darnell <notifications@github.com> wrote:

> [image: :lgtm:]
> <https://camo.githubusercontent.com/41b3ec3419116e3b3f84507ad965646ce4ce1f0c/68747470733a2f2f72657669657761626c652e696f2f6c67746d2e706e67>
> ------------------------------
>
> Reviewed 10 of 10 files at r3.
> Review status: all files reviewed at latest revision, 3 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *pkg/storage/replica_proposal.go, line 520 at r3
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KngiUQxwO1qAQ6mhxAm:-KngiUQxwO1qAQ6mhxAn:b-d6gbz2>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/1e5ca533aceaa64a642dc1ce2644b3642ba779b8/pkg/storage/replica_proposal.go#L520>):*
>
> 			panic(err)
> 		} else {
> 			log.Warningf(ctx, "file %s already exists, likely doubly applied Raft command; skipping", path)
>
> The file could have been incompletely written by an earlier version of the
> process. We should verify that the existing file has the contents we want
> to write.
> ------------------------------
>
> *pkg/storage/replica_sideload_test.go, line 48 at r2
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBdLTIjVt7eGFiL873:-KnghYyupykmPXZIc1un:b-b8kbcr>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload_test.go#L48>):*
> *Previously, tschottdorf (Tobias Schottdorf) wrote…*
>
> because it's nice to be able to run TESTS=Sideloading and get the
> sideloading-related tests. But maybe I went overboard with this. The
> leaktest thing is unfortunate, but more a limitation of leaktest than
> indicating that this isn't a good pattern. Happy to make these individual
> tests if you think that's what they should be.
>
> TESTS=Sideloading does a regex search on the test names, so as long as
> all the tests have Sideloading in their names it will work the way you
> want.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-:-KngjLuqrE1oDK5tL2fs:bnfp4nl>*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/16509#issuecomment-311546652>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135OcVokw0ySl3i_1moScCl5grHW1wks5sIcg5gaJpZM4N5JWb>
> .
>
-- 

-- Tobias

tschottdorf commented:
Ok, back with more info and a better understanding.

The root cause of the panics was the use of RocksDB range deletion tombstones. These are a known problem, but somehow the change to sideloading massively amplified the problem.

The TL;DR is:

* with ranged tombstones enabled, most nodes in a typical fresh cluster have a range deletion tombstone that covers pretty much "everything" (since we use one during snapshot application, and the initial range spans "everything").
* with our RocksDB version on `master` (at the time of writing), ingestion of SSTables has undefined behaviour w.r.t range tombstones (i.e. it what you see depends on at which levels each are).
* sideloading massively reduces how much data passes through RocksDB, and so (this is theory) the deletion tombstones stuck around for much longer, or at least remained at higher levels, so that ingestions were likely to become randomly invisible, which in turn causes inconsistency.

I added an assertion around the ingestion that checks the number of keys before and after, and I observed that for ranges failing consistency, at least some of the members ingested an SSTable without gaining any keys (i.e. the ingestion was invisible), in line with what you expect if a range tombstone and an ingestion happened with wrong precedence.

Dan provided me with an updated RocksDB version (where this is fixed) and the problem goes away.

I also tried directly repro'ing in a test, but it depends on internal RocksDB LSM state; didn't manage. Confident it *could* be done, but it doesn't seem necessary with the current understanding.

We're going to have to disable range clears on `master` whenever testing restore, but other than that we should be good to go.

Don't re-review just yet, I have more updates to clean up and push.

tschottdorf commented:
@danhhz, can you take a look, in particular at the last commit? I was thinking about/testing the delete-before-ingest logic taking into account that we also want to ingest-before-commit. That creates the awkward situation in which the batch contains deletions for the same keys that we also ingest, and so the order in which operations apply matters -- I thought the right thing to do would be to open the RocksDB batch *first*, then ingest, then commit the batch. But that seems to have the same outcome as ingest - create batch - commit, namely the ingest goes AWOL... Any thoughts?

```go
		var usuallyNilBatch engine.Batch
		// AddSSTable ingestions run before the actual batch. This makes sure that
		// when the Raft command is applied, the ingestion has definitely succeeded.
		if raftCmd.ReplicatedEvalResult.AddSSTable != nil {
			// Create the batch now so that it is open before the ingestion.
			// Otherwise, if the batch contains a deletion for a (RocksDB-level)
			// key present in the import, and the batch applies *on top of* the
			// ingestion, it will nuke the ingested key. So we want the
			// following order:
			// - batch is created
			// - SSTable is ingested
			// - deletions are written to batch
			// - batch commits.
			//
			// TODO(tschottdorf): this doesn't actually work. Find out why and
			// how we can fix it: TestRaftSSTableSideloadingProposal/empty=true
			// fails; after the second proposal, the key is missing. It is fixed
			// when we move applyRaftCommand before this block.
			//
			// TODO(tschottdorf): refactor for beauty. We could always pass the
			// batch into applyRaftCommand.
			usuallyNilBatch = r.store.engine.NewWriteOnlyBatch()
			addSSTablePreApply(
				ctx,
				r.store.engine,
				r.raftMu.sideloaded,
				term,
				index,
				raftCmd.ReplicatedEvalResult.StartKey,
				raftCmd.ReplicatedEvalResult.EndKey,
				*raftCmd.ReplicatedEvalResult.AddSSTable,
			)
			raftCmd.ReplicatedEvalResult.AddSSTable = nil
		}

		raftCmd.ReplicatedEvalResult.Delta, pErr = r.applyRaftCommand(
			ctx, idKey, raftCmd.ReplicatedEvalResult, writeBatch, usuallyNilBatch)
```

tschottdorf commented:
Ready for another look. Since @danhhz is out this week, I'd be glad if you could do the honors, @bdarnell.

I addressed the problem of clobbering the own ingestion by falling back to doing simple WriteBatch (i.e. no ingestion) when there is existing data. This may not be perfect but is less sketchy than other solutions would likely end up being (and after all, we rarely expect there to be data in our key range except during reproposals, etc).

---

Review status: 5 of 21 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_proposal.go, line 520 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KngiUQxwO1qAQ6mhxAm:-KoJPtF2It0exUpXvVHC:bs4pg9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e5ca533aceaa64a642dc1ce2644b3642ba779b8/pkg/storage/replica_proposal.go#L520)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The file could have been incompletely written by an earlier version of the process. We should verify that the existing file has the contents we want to write. 
</blockquote></details>

Comment is outdated.

---

*[pkg/storage/replica_sideload_test.go, line 48 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-KnBdLTIjVt7eGFiL873:-KoJPxKPjrKJgItZ9wNa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58fe962efd50122c6640ed5c36e50becc9a159/pkg/storage/replica_sideload_test.go#L48)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`TESTS=Sideloading` does a regex search on the test names, so as long as all the tests have `Sideloading` in their names it will work the way you want. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-:-KoJQ0aq-669wG6XyCRW:b-gh19bh)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: Raft SSTable sideloading

Implement the Raft SSTable Sideloading RFC. Briefly put,

- on the Replica proposal path, detect AddSSTable commands and mark them as
  Raft command version v2.
- Intercept such entries on their way into the Raft log, and store the
  actual SSTable on disk.
- On the wire, always send the full ("inlined") proposals, including in
  snapshots.

See #16263.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 18 of 18 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16509#-:-KoJpjw-XC9zWvaA1YK4:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTRs, @bdarnell and @danhhz! Time to get this show on the road.
